### PR TITLE
Apply & bundle edits for non-message events.

### LINKED ISSUE #12503: Synapse does not bundle `m.replace` aggregations for events other than `m.room.message`
Per MSC2675, when the server returns an event, it is supposed to bundle the aggregation of any related events into the response.

This includes `m.replace` relations: for an `m.room.message` event, `/context/$original_event_id` might return:

```json5
{
  "event": {
    ...
    "event_id": "$original_event_id",
    "type": "m.room.message",
    "unsigned": {
      "m.relations": {
        "m.replace": {
          "event_id": "$replacement_event_id",
          "origin_server_ts": 1650390324643,
          "sender": "@richvdh:matrix.org"
        }
      }
    },
    "user_id": "@richvdh:matrix.org"
  }
}
```

However, for other event types, no `m.replace` aggregation is returned.

It's not as if `m.room.message` is the only event type you can edit (consider, for example, encrypted events), so this seems odd.

### LINKED ISSUE #12793: Synapse does not apply edits to events other than `m.room.message`
MSC2676 [says](https://github.com/uhoreg/matrix-doc/blob/aggregations-edits/proposals/2676-message-editing.md#server-behaviour):

> Whenever a homeserver would return an event via the Client-Server API, it should check for any valid, applicable `m.replace` event, and if one is found, it should first modify the content of the original event according to the `m.new_content` of the most recent edit.

It also [gives](https://github.com/uhoreg/matrix-doc/blob/aggregations-edits/proposals/2676-message-editing.md#validity-of-message-edit-events) a comprehensive list of what comprises a "valid" edit event.

Synapse does this, but only where the events are of type `m.room.message`. This is incorrect: it should also do so for other event types. Examples include polls ([`org.matrix.msc3381.poll.start`](https://github.com/matrix-org/matrix-spec-proposals/pull/3381)), which can be edited via the Element UI, and stickers ([`m.sticker`](https://spec.matrix.org/v1.2/client-server-api/#sticker-messages)), which cannot be edited via the UI but can via the API.

This is closely related to #12503, but that concerns aggregations, whereas this issue is about updating the `content` of the original event.

### PR DESCRIPTION
**Pull request by clokep**
Currently Synapse only fetches edits for `m.room.message` events, this is not correct according to the specification. It should [apply for all types of events](https://spec.matrix.org/v1.4/client-server-api/#validity-of-replacement-events).

This manifest in two related bugs (see below):

* No edit information is bundled for events which aren't `m.room.message`.
* `m.new_content` is not applied where available.

Due to the structure of our code both of these have the same room cause, which is simply that we enforce that the event type must be `m.room.message` when fetching edits from the database. We relax that in this PR.

Note that I didn't modify any tests since the current tests send `m.room.message` events, which is still valid.

Fixes #12793, fixes #12503

### CODE REVIEW

File: synapse/storage/databases/main/relations.py [hunk 1]
```diff
@@ -443,6 +440,8 @@ def _get_applicable_edits_txn(txn: LoggingTransaction) -> Dict[str, str]:
 
         edits = await self.get_events(edit_ids.values())  # type: ignore[attr-defined]
 
+        # TODO Should this assert m.new_content exists in the content for cleartext events?
+
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/storage/databases/main/relations.py:444]: Oh hm, I think I left this as a note to myself. If there's no `m.new_content` than Synapse will replace with empty content:

https://github.com/matrix-org/synapse/blob/d6ae14e60e8c5316471384cc52502364edaea5fd/synapse/events/utils.py#L459

The weird thing is that the spec says (emphasis mine):

> The replacement event (**once decrypted**, if appropriate) must have an `m.new_content` property.

So in the case of an encrypted event it *will not* have an `m.new_content` (and Synapse will essentially nuke the events content when returning it), clients could then fetch the replacement event, decrypt it and apply the content?

tl;dr for this comment is that it should get removed, I think. The spec about [server behavior](https://spec.matrix.org/latest/client-server-api/#server-side-replacement-of-content) should be clarified though.

[the remark this answers is not in the document]
Comment (clokep) [reply synapse/storage/databases/main/relations.py:444]: @richvdh Does the above sound like Synapse is doing the correct thing there or should it not be leaving the original content alone if there's no `m.new_content` field? (Also what happens if it is not a dict, etc. etc.)

[the remark this answers is not in the document]
Comment (clokep) [reply synapse/storage/databases/main/relations.py]: I'm not going to block on this and will file a separate issue.
